src/hotspot/share/opto/node.cpp

```diff
@@ -2208,25 +2202,22 @@ bool PrintBFS::parse_options() {
   return true;
 }
 
-bool PrintBFS::filter_category(Node* n, Filter& filter) {
+bool PrintBFS::filter_category(const Node* n, Filter& filter) {
   const Type* t = n->bottom_type();
-  switch (t->category()) {
-    case Type::Category::Data:
-      return filter._data;
-    case Type::Category::Memory:
-      return filter._memory;
-    case Type::Category::Mixed:
-      return filter._mixed;
-    case Type::Category::Control:
-      return filter._control;
-    case Type::Category::Other:
-      return filter._other;
-    case Type::Category::Undef:
-      n->dump();
-      assert(false, "category undef ??");
-    default:
-      n->dump();
-      assert(false, "not covered");
+  if (filter._data && t->has_category(Type::Category::Data)) {
+    return true;
+  }
+  if (filter._memory && t->has_category(Type::Category::Memory)) {
+    return true;
+  }
+  if (filter._mixed && t->has_category(Type::Category::Mixed)) {
+    return true;
+  }
+  if (filter._control && (t->has_category(Type::Category::Control) || n->is_CFG())) {
+    return true;
+  }
+  if (filter._other && t->has_category(Type::Category::Other)) {
+    return true;
```

Comment:
Just a suggestion: To make it clear that you are only special casing `Mixed` you could leave the `switch` statement and only do the additional checks for `Mixed`. Since this category check is specific to the filtering of `dump_bfs()` and not per-se something you normally perform on a type, I suggest to move this function to the `Filter` class (if that's possible). This would also require to change the implementation of `has_category()` - if it's too complicated, just leave it as it is. It's fine like that.

Reply:
I now made a single if statement out of it, and moved it to `Filter`. It was not evident for me how to use `switch,` because both the filter and the node can have multiple categories.